### ISSUE TITLE
False positives in "prefer-const"

### ISSUE THREAD

**Issue by bartlomieju** [denoland/deno_lint#358]
```
(prefer-const) 'allowBlockScalars' is never reassigned. Use 'const' instead
  let allowBlockScalars: boolean,
      ^^^^^^^^^^^^^^^^^
    at /Users/biwanczuk/dev/deno/std/encoding/_yaml/loader/loader.ts:1442:6

(prefer-const) 'modulePaths' is never reassigned. Use 'const' instead
let modulePaths: string[] = [];
    ^^^^^^^^^^^
    at /Users/biwanczuk/dev/deno/std/node/module.ts:628:4
```

CC @magurotuna 

I temporarly ignored them in `std/` to upgrade the linter, but we should fix those problems.

**Comment by magurotuna** [comment]
I'll look into it.

**Comment by magurotuna** [comment]
Seems like the second one, `std/node/module.ts`, has something to do with hoisting... I didn't care about that. I may need some time to fix it.

**Comment by bartlomieju** [comment]
> Seems like the second one, `std/node/module.ts`, has something to do with hoisting... I didn't care about that. I may need some time to fix it.

That's okay, I will just disable `prefer-const` rule for this release

**Comment by magurotuna** [comment]
```
 262   │ (prefer-const) 'openStringSyntax' is never reassigned. Use 'const' instead
 263   │   --> /home/yusuktan/Repos/github.com/magurotuna/deno/std/encoding/toml.ts:74:8
 264   │    |
 265   │ 74 |     let openStringSyntax = "";
 266   │    |         ^^^^^^^^^^^^^^^^
 267   │    |
```

-> hoisting

```
 346   │ (prefer-const) 'modulePaths' is never reassigned. Use 'const' instead
 347   │    --> /home/yusuktan/Repos/github.com/magurotuna/deno/std/node/module.ts:628:4
 348   │     |
 349   │ 628 | let modulePaths: string[] = [];
 350   │     |     ^^^^^^^^^^^
 351   │     |
```

-> hoisting

```
 605   │ (prefer-const) 'allowBlockScalars' is never reassigned. Use 'const' instead
 606   │     --> /home/yusuktan/Repos/github.com/magurotuna/deno/std/encoding/_yaml/loader/loader.ts:1442:6
 607   │      |
 608   │ 1442 |   let allowBlockScalars: boolean,
 609   │      |       ^^^^^^^^^^^^^^^^^
 610   │      |
```

-> maybe nested assignment expression

-----------------------

So there are two causes why false positives occur. I'll fix one by one.

**Comment by magurotuna** [comment]
The following might also be an edge case of `prefer-const`. It has a `for-in` statement of which the variable is declared beforehand, like `let pairKey = "";`

From `std/encoding/_yaml/type/omap.ts`:

```ts
function resolveYamlOmap(data: Any): boolean {
  const objectKeys: string[] = [];
  let pairKey = "";
  let pairHasKey = false;

  for (const pair of data) {
    pairHasKey = false;

    if (_toString.call(pair) !== "[object Object]") return false;

    for (pairKey in pair) {
      if (_hasOwnProperty.call(pair, pairKey)) {
        if (!pairHasKey) pairHasKey = true;
        else return false;
      }
    }

    if (!pairHasKey) return false;

    if (objectKeys.indexOf(pairKey) === -1) objectKeys.push(pairKey);
    else return false;
  }

  return true;
}
```

**Comment by bartlomieju** [comment]
@magurotuna is it marked as a violation? If so, then it's definitely a false positive ;)

**Comment by magurotuna** [comment]
@bartlomieju currently not, but in my latest refactoring (#362), the above case may become an edge case, so I commented as a memo. I'm addressing this right now :)
